### ISSUE TITLE
Treat each match case as its own scope

### ISSUE THREAD

**Issue by charliermarsh** [charliermarsh/ruff#3099]
Given:

```py
def f():
    match Car("Toyota", "Corolla"):
        case Car("Toyota", model):
            print(model)
        case Car(make, "Corolla"):
            print(make)
        case Car(_, _):
            print(make, model)
```

The `(make, model)` in the final case should be flagged as undefined.

**Comment by charliermarsh** [comment]
It's actually a little more narrow than this: variables defined in other `match` patterns shouldn't be accessible inside the `match` body (but any variable defined in any `match` pattern or `match` body should be accessible from the defining function).

**Comment by charliermarsh** [comment]
This is kind of the same as variables defined within if-branches etc.
